File: workers/worker.js
const { delay, getRandom, colorize } = require("./utils")
require("dotenv").config()

const executeTask = (num1, num2, type) => {
  switch (type) {
    case "addition":
      return num1 + num2
      break
    case "subtraction":
      return num1 - num2
      break
    case "multiplication":
      return num1 * num2
      break
    case "division":
      return num1 / num2
      break
    default:
      throw new Error("Unsupported operation type: " + type)
  }
}

const getNextTask = async ({ queue, type }) => {
  try {
    await delay(getRandom(5, 7, "worker") * 1000)

    const requestBody = queue ? { queue } : { type }

    const response = await fetch(`http://127.0.0.1:8383/get-available-tasks`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify(requestBody),
    })

    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`)
    }

    const data = await response.json()

    return data
  } catch (error) {
    return null
  }
}

const sendResults = async (id, result, endTime, error) => {
  try {
    const response = await fetch(`http://127.0.0.1:8383/submit-results`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
      },
      body: JSON.stringify({ id, endTime, result, error }),
    })
    return response
  } catch (error) {
    console.error("Error while sending result:", error.message)
  }
}

const runWorker = async () => {
  while (true) {
    const type = "addition"

    try {
      console.log()
      console.log(colorize(`\n Fetching tasks...`, "magenta"))
      await delay(getRandom(5, 7, "worker") * 1000)

      const response = await getNextTask({ type })

      if (response === null) {
        console.log(
          colorize(` No tasks found, worker going sleep mode`, "yellow")
        )
        await delay(10000)
        continue
      }
      await delay(getRandom(4, 5, "worker") * 1000)

      console.log(colorize(` ↳ Task found`, "cyan"))
      console.log(
        colorize(` ↳ Task details: ${JSON.stringify(response)}`, "cyan")
      )
      const { id, params } = response
      const { num1, num2 } = params

      const result = executeTask(num1, num2, type)

      console.log(colorize(` ↳ Task completed successfully`, "green"))
      if (result) {
        await sendResults(id, result, null)
        console.log(colorize(` ↳ Results submitted successfully`, "green"))
      } else if (error) {
        await sendResults(id, result, error)
        console.log(colorize(` ↳ Results submitted successfully`, "green"))
      } else {
        console.log(
          colorize(` ↳ Error sending results: ${error.message}`, "red")
        )
      }
    } catch (error) {
      console.error("Error in runWorker:", error.message)
    }
  }
}

runWorker()

File: workers/python/worker.py
import time
import json
import requests
from random import randint

root = "http://127.0.0.1:8383"

halt_n_execute = False  # make it False or lighting fast execution ;)


def execute_task(num1, num2, operation_type):
    if operation_type == "add":
        return num1 + num2
    elif operation_type == "subtract":
        return num1 - num2
    elif operation_type == "multiply":
        return num1 * num2
    elif operation_type == "divide":
        return num1 / num2
    else:
        raise ValueError("Unsupported operation type: " + operation_type)


def get_next_task(queue=None, operation_type=None):
    try:
        request_body = {"queue": queue} if queue else {"type": operation_type}
        response = requests.post(
            root + "/get-next-available-task", json=request_body)

        data = response.json()
        return data
    except Exception as error:
        return None


def send_results(task_id, result, error):
    try:
        response = requests.post(
            root + "/submit-results",
            json={"id": task_id, "result": result, "error": error},
        )
        return response
    except Exception as error:
        print("Error while sending result:", error)


def run_worker():
    valid_operation_types = ["add", "subtract", "multiply", "divide"]
    operation_type = input(
        'Type & enter operation type ["add", "subtract", "multiply", "divide"] : ')

    while operation_type.strip() not in valid_operation_types:
        print("Invalid operation type. Please enter a valid operation type.")
        operation_type = input("Type & enter operation type:")

    while True:
        try:
            print(f"\nFetching tasks...")
            if halt_n_execute:
                input("Press Enter to start processing...")

            response = get_next_task(operation_type=operation_type)
            if isinstance(response, dict) and "message" in response:
                print("\nNo tasks found, worker going to sleep mode")
                time.sleep(20)
                continue

            print(f"Task found\nTask details: {json.dumps(response)}")

            task_id, params = response["id"], response["params"]
            num1, num2 = params["num1"], params["num2"]
            if halt_n_execute:
                input("Press Enter to execute...")
            result = execute_task(num1, num2, operation_type)
            print(result)

            print("Task completed successfully")

            if halt_n_execute:
                input("Press Enter to submit...")

            if (result != 0):
                send_results(task_id, result, None)
                print("Results submitted successfully")
            elif (result == 0):
                error = "result is zero"
                send_results(task_id, result, error)
                print(f"Error sending results: {error}")

        except Exception as error:
            print("Error in run_worker:", error)


if __name__ == "__main__":
    run_worker()
